src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotDefineFinalizersForTypesDerivedFromMemoryManagerTests.cs

```diff
@@ -0,0 +1,223 @@
+
+            namespace TestNamespace
+            {
+                class TestClass
+                {
+                    private void TestMethod() { }
+                }
+            }");
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+            Imports System
+
+            Namespace TestNamespace
+                Class TestClass
+                    Private Sub TestMethod()
+                    End Sub
+                End Class
+            End Namespace");
+        }
+
+        [Fact]
+        public async Task ClassHavingFinalizerButNotDerivedFromMemoryManagerOK()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+            using System;
+
+            namespace TestNamespace
+            {
+                class TestClass
+                {
+                    private void TestMethod()
+                    {
+                    }
+
+                    ~TestClass() 
+                    {
+                        TestMethod();
+                    }
+                }
+            }");
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+            Imports System
+
+            Namespace TestNamespace
+                Class TestClass
+                    Private Sub TestMethod()
+                    End Sub
+
+                    Protected Overrides Sub Finalize()
+                        TestMethod()
+                    End Sub
+                End Class
+            End Namespace");
+        }
+
+        [Fact]
+        public async Task ClassDerivedFromMemoryManagerNoFinilizerOK()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+            using System;
+            using System.Buffers;
+
+            namespace TestNamespace
+            {
+                abstract class TestClass<T> : MemoryManager<T>
+                {
+                }
+            }");
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+            Imports System
+            Imports System.Buffers
+
+            Namespace TestNamespace
+                MustInherit Class TestClass(Of T)
+                    Inherits MemoryManager(Of T)
+                End Class
+            End Namespace
+            ");
+        }
+
+        [Fact]
+        public async Task ClassDerivedFromMemoryManagerWithFinilizerWarns()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+            using System;
+            using System.Buffers;
+
+             namespace TestNamespace
+             {
+                class TestClass<T> : MemoryManager<T>
+                {
+                    public override Span<T> GetSpan()
+                    {
+                        throw new NotImplementedException();
+                    }
+
+                    public override MemoryHandle Pin(int elementIndex = 0)
+                    {
+                        throw new NotImplementedException();
+                    }
+
+                    public override void Unpin() { }
+
+                    ~TestClass() { }
```

Comment:
📝 Markup syntax looks like this:

```suggestion
                    [|~TestClass|]() { }
```

Reply:
@sharwell Does this markup syntax have the effect of the `VerifyAnalyzerAsync` method automatically checking where the warning would be rendered?

Reply:
Thanks @sharwell, I am wondering about what it does exactly too, as its named markup i guess it just give visual effect for where the errors should reported

Reply:
Yes, the test will fail if the analyzer does not report a diagnostic at the annotated markup span.

Reply:
NOTE: If you have multiple different diagnostic IDs being reported by the analyzer, you can use the markup syntax with prefixed diagnostic ID, like validating `RS2008` below: https://github.com/dotnet/roslyn-analyzers/blob/d02078fe8f2a1051d41b9c5cd6899b2ef6d41649/src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/ReleaseTrackingAnalyzerTests.cs#L71

For analyzers that report a single diagnostic ID, the syntax that Sam provided implicitly uses that diagnostic ID for validation.

Reply:
Yes, it verifies that the diagnostic is reported exactly on the indicated span (as opposed to just requiring that it is within the span.

* If another diagnostic is reported elsewhere, the test will fail
* If the diagnostic is _not_ reported at the marked span, the test will fail
* If the diagnostic does not start _at_ the start of the span, or does not end _at_ the end of the span, the test will fail

Reply:
> Yes, it verifies that the diagnostic is reported exactly on the indicated span (as opposed to just requiring that it is within the span.
> 
> If another diagnostic is reported elsewhere, the test will fail
> If the diagnostic is not reported at the marked span, the test will fail
> If the diagnostic does not start at the start of the span, or does not end at the end of the span, the test will fail

Great, thanks, then I can remove  `.WithSpan(9, 45, 9, 53));`

Reply:
Am I missing something, just adding `[||]` around finalizer not working. `[|~TestClass|]() { }`
From error I'm guessing its expecting an argument for the message or so, my warning message not expecting any argument though 
The error:
```
   Microsoft.CodeAnalysis.Testing.Verifiers.EqualWithMessageException : Context: Diagnostics of test state\r\nMismatch between number of diagnostics returned, expected \"2\" actual \"1\"\r\n\r\nDiagnostics:\r\n// /0/Test0.cs(21,22): warning CA2015: Adding a finalizer to a type derived from MemoryManager<T> may permit memory to be freed while it is still in use by a Span<T>.\r\nVerifyCS.Diagnostic().WithSpan(21, 22, 21, 31).WithArguments(\"Finalize\"),\r\n\r\n\r\nAssert.Equal() Failure\r\nExpected: 2\r\nActual:   1
   at Microsoft.CodeAnalysis.Testing.Verifiers.XUnitVerifier.Equal[T](T expected, T actual, String message) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Testing.Verifiers.XUnit/XUnitVerifier.cs:line 51
```

Reply:
did you remove the `VerifyCS.Diagnostic(...)` part? If not, it is expecting the diagnostic message twice (once because of the markup, and a second instance because of the manual parameter)

Reply:
Oh really, no I only removed `.WithSpan(9, 45, 9, 53)` 😛